lib/expr.ml

```diff
@@ -1,46 +1,65 @@
 open Ty
 
-type t =
-  { e : expr
-  ; ty : Ty.t
-  }
+module M = struct
+  type t =
+    { e : expr
+    ; ty : Ty.t
+    }
 
-and expr =
-  | Val of Value.t
-  | Ptr of int32 * t
-  | Unop of unop * t
-  | Binop of binop * t * t
-  | Triop of triop * t * t * t
-  | Relop of relop * t * t
-  | Cvtop of cvtop * t
-  | Symbol of Symbol.t
-  | Extract of t * int * int
-  | Concat of t * t
+  and expr =
+    | Val of Value.t
+    | Ptr of int32 * t
+    | Unop of unop * t
+    | Binop of binop * t * t
+    | Triop of triop * t * t * t
+    | Relop of relop * t * t
+    | Cvtop of cvtop * t
+    | Symbol of Symbol.t
+    | Extract of t * int * int
+    | Concat of t * t
 
-let ( @: ) e ty = { e; ty }
+  let equal (e1 : t) (e2 : t) : bool =
+    if not (e1.ty = e2.ty) then false
+    else
+      match (e1.e, e2.e) with
+      | Val v1, Val v2 -> Value.equal v1 v2
+      | Ptr (b1, o1), Ptr (b2, o2) -> b1 = b2 && o1 == o2
+      | Unop (op1, e1), Unop (op2, e2) -> op1 = op2 && e1 == e2
+      | Cvtop (op1, e1), Cvtop (op2, e2) -> op1 = op2 && e1 == e2
+      | Binop (op1, e1, e3), Binop (op2, e2, e4) ->
+        op1 = op2 && e1 == e2 && e3 == e4
+      | Relop (op1, e1, e3), Relop (op2, e2, e4) ->
+        op1 = op2 && e1 == e2 && e3 == e4
+      | Triop (op1, e1, e3, e5), Triop (op2, e2, e4, e6) ->
+        op1 = op2 && e1 == e2 && e3 == e4 && e5 == e6
+      | Symbol s1, Symbol s2 -> Symbol.equal s1 s2
+      | Extract (e1, h1, l1), Extract (e2, h2, l2) ->
+        e1 == e2 && h1 = h2 && l1 = l2
+      | Concat (e1, e3), Concat (e2, e4) -> e1 == e2 && e3 == e4
+      | _ -> false
+
+  let hash = Hashtbl.hash
+end
+
+include M
+module H = Hashtbl.Make (M)
```

Comment:
It's probably better to use `Ephemeron.K1.Make`. At the time of that paper, ephemerons were not part of the stdlib, so they had to redo it from scratch, but now it's easy. :)

Reply:
Will do! Thanks for the suggestion :smile: